### DISCUSSION TITLE
Dagster DBT Asset Key Naming

### DISCUSSION THREAD

**Discussion by jackwillisupside** [dagster-io/dagster#19241] [Q&A] [answered]
### What's the use case?

We're spinning up a data-mesh like dbt/python instance and are hitting issues where the name of a model in code location 1 is the same as the name of a model in code location 2 (think finance.staging.user and marketing.staging.user). We'd love if the asset key for dbt models were code location, schema, table rather than just schema and table as it requires either 1. specifying through meta the asset key in the schema.yml or 2. enforcing our users (who might not know what other users are doing) to have distinct names across our instance.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by jackwillisupside** [comment]
We tried to implement the custom asset key method and adding a prefix of the name of the code location to our dbt asset keys -- but can't seem to get it to work across code locations. When specifying a custom key in the meta: in dbt schema, it thinks that the source is controlled by the "consumer" code location rather than the actual dagster asset in the "producer" code location

**Comment by jackwillisupside** [comment]
We found a work around using the dbt metadata and a custom get_asset_key to get this to work, so we're fine for the time being.

**Comment by jzamanxometry** [comment]
@jackwillisupside could you post any code that you used to get it to work, my team and I would be very interested

**Comment by jackwillisupside** [comment, marked as answer]
For context, subsystem name is the name of our code locations and for consuming code locations (specifically sources) we require the folder that the sources.yml to live in to match the name of the producing subsystem (i.e. if im grabbing something from the finance code location, the source folder would be named contracts_finance,src_contracts_finance, or just finance).

```
 @classmethod
    def get_asset_key(self, dbt_resource_props: Mapping[str, Any]) -> AssetKey:
        asset_key = super().get_asset_key(dbt_resource_props)

        if dbt_resource_props["unrendered_config"].get('meta', {}).get('dagster', {}).get('asset_key'):
            pass
        elif dbt_resource_props["resource_type"] != "source":
            # if the asset is not a source, add the subsystem name as a prefix
            asset_key = asset_key.with_prefix(os.getenv('SUBSYSTEM_NAME').replace('-', '_'))
        else:
            # if the asset is a source and is not, add the source subsystem as a prefix
            source_system = dbt_resource_props["original_file_path"].split('/')[2]
            source_system = source_system.replace('contracts_', '').replace('src_', '').replace('-', '_')
            asset_key = asset_key.with_prefix(source_system)
        return asset_key
```

**Comment by jackwillisupside** [comment]
Hope this is helpful @jzamanxometry !!

**Comment by jzamanxometry** [comment]
Some clarification questions
1. Did you use the default dagster-dbt setup for this?
1a. If so, where in dagster dbt project did you overwrite/define this custom get_asset_key and how did you use it? 
1b. if not, still curious how and where you overrode this for asset selection

> For context, subsystem name is the name of our code locations and for consuming code locations (specifically sources) we require the folder that the sources.yml to live in to match the name of the producing subsystem (i.e. if im grabbing something from the finance code location, the source folder would be named contracts_finance,src_contracts_finance, or just finance).
> 
> ```
>  @classmethod
>     def get_asset_key(self, dbt_resource_props: Mapping[str, Any]) -> AssetKey:
>         asset_key = super().get_asset_key(dbt_resource_props)
> 
>         if dbt_resource_props["unrendered_config"].get('meta', {}).get('dagster', {}).get('asset_key'):
>             pass
>         elif dbt_resource_props["resource_type"] != "source":
>             # if the asset is not a source, add the subsystem name as a prefix
>             asset_key = asset_key.with_prefix(os.getenv('SUBSYSTEM_NAME').replace('-', '_'))
>         else:
>             # if the asset is a source and is not, add the source subsystem as a prefix
>             source_system = dbt_resource_props["original_file_path"].split('/')[2]
>             source_system = source_system.replace('contracts_', '').replace('src_', '').replace('-', '_')
>             asset_key = asset_key.with_prefix(source_system)
>         return asset_key
> ```

**Comment by jackwillisupside** [comment]
@jzamanxometry we started with the standard setup and have moved to a custom dbt translator in order to add some metadata and group names. So it looks like this in our locations.

```
class CustomDagsterDbtTranslator(DagsterDbtTranslator):

    @classmethod
    def get_asset_key(self, dbt_resource_props: Mapping[str, Any]) -> AssetKey:
        asset_key = super().get_asset_key(dbt_resource_props)

        if dbt_resource_props["unrendered_config"].get('meta', {}).get('dagster', {}).get('asset_key'):
            pass
        elif dbt_resource_props["resource_type"] != "source":
            # if the asset is not a source, add the subsystem name as a prefix
            asset_key = asset_key.with_prefix(os.getenv('SUBSYSTEM_NAME').replace('-', '_'))
        else:
            # if the asset is a source and is not, add the source subsystem as a prefix
            source_system = dbt_resource_props["original_file_path"].split('/')[2]
            source_system = source_system.replace('contracts_', '').replace('src_', '').replace('-', '_')
            asset_key = asset_key.with_prefix(source_system)
        return asset_key

# Set up project Definitions
dbt_project_path = file_relative_path(__file__, "../../dbt_code_location")
manifest=get_manifest(dbt_project_path,dbt_manifest_path)


#Load definitions into dbt assets for dagster
@dbt_assets(
    manifest=manifest,
    dagster_dbt_translator=CustomDagsterDbtTranslator(),
)
```

**Comment by jzamanxometry** [comment]
That's the context I was missing. Thanks @jackwillisupside . This is super helpful as we are planning to migrate some of our project to do this and will be building inter-dbt project lineage.
